Source/WebCore/dom/Document.cpp

```diff
@@ -6215,8 +6215,38 @@ void Document::setCookieURL(const URL& url)
 {
     if (m_cookieURL == url)
         return;
+
     m_cookieURL = url;
     invalidateDOMCookieCache();
+    updateCachedCookiesEnabled();
+}
+
+void Document::setFirstPartyForCookies(const URL& url)
+{
+    if (m_firstPartyForCookies == url)
+        return;
+
+    m_firstPartyForCookies = url;
+    updateCachedCookiesEnabled();
+}
+
+void Document::setCachedCookiesEnabled(bool enabled)
+{
+    m_cachedCookiesEnabled = enabled;
+}
+
+void Document::updateCachedCookiesEnabled()
+{
+    RefPtr page = this->page();
+    if (!page || !page->settings().cookieEnabled() || canAccessResource(ScriptExecutionContext::ResourceType::Cookies) == ScriptExecutionContext::HasResourceAccess::No) {
+        setCachedCookiesEnabled(false);
+        return;
+    }
+
+    page->cookieJar().cookiesEnabledAsync(*this, [weakThis = WeakPtr<Document, WeakPtrImplWithEventTargetData> { *this }](bool enabled) mutable {
```

Comment:
`weakDocument = WeakPtr { *this }` (no need for template parameters and renamed)